blueocean-rest-impl/src/main/java/io/jenkins/blueocean/service/embedded/rest/ArtifactContainerImpl.java

```diff
@@ -1,122 +1,45 @@
 import com.google.common.collect.Iterators;
-import com.google.common.util.concurrent.Service;
-import hudson.Functions;
 import hudson.model.Run;
-import io.jenkins.blueocean.commons.ServiceException;
 import io.jenkins.blueocean.rest.Reachable;
+import io.jenkins.blueocean.rest.factory.BlueArtifactFactory;
 import io.jenkins.blueocean.rest.hal.Link;
 import io.jenkins.blueocean.rest.model.BlueArtifact;
 import io.jenkins.blueocean.rest.model.BlueArtifactContainer;
-import io.jenkins.blueocean.rest.model.Resource;
-import jenkins.util.VirtualFile;
-import org.kohsuke.stapler.Stapler;
 
+import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
-import java.io.IOException;
 import java.util.Iterator;
-import java.util.List;
 
 public class ArtifactContainerImpl extends BlueArtifactContainer {
     final private Run run;
     final private Link self;
+
     public ArtifactContainerImpl(Run r, Reachable parent) {
         this.run = r;
         this.self = parent.getLink().rel("artifacts");
     }
 
-
     @Override
     public Link getLink() {
         return self;
     }
 
     @Override
     public BlueArtifact get(final String name) {
-        // Check security for artifacts
-        if(Functions.isArtifactsPermissionEnabled() && !run.hasPermission(Run.ARTIFACTS)) {
-            return null;
-        }
-
-        final VirtualFile file = run.getArtifactManager().root().child(name);
-        try {
-            if(file == null || !file.exists() || !file.isFile()) {
-                return null;
-            }
-        } catch (IOException e) {
-            throw new ServiceException.UnexpectedErrorException("Something is wrong with an artifact",e);
-        }
-        return new BlueArtifact() {
-            @Override
-            public String getName() {
-                return file.getName();
-            }
-
-            @Override
-            public String getPath() {
-                return name;
-            }
-
-            @Override
-            public String getUrl() {
-                return Stapler.getCurrentRequest().getContextPath() +
-                    "/" + run.getUrl()+"artifact/"+ name;
-            }
-
-            @Override
-            public long getSize() {
-                try {
-                    return file.length();
-                } catch (IOException e) {
-                    throw new ServiceException.UnexpectedErrorException("qError getting file length", e);
-                }
-            }
-
-            @Override
-            public Link getLink() {
-                return new Link(getUrl());
-            }
-        };
-    }
-
-    @Override
-    public Iterator<BlueArtifact> iterator(int start, int limit) {
-        // Check security for artifacts
-        if(Functions.isArtifactsPermissionEnabled() && !run.hasPermission(Run.ARTIFACTS)) {
-           return Iterators.emptyIterator();
-        }
-
-        List<Run.Artifact> artifactsUpTo = run.getArtifactsUpTo(limit);
```

Comment:
Ah, I missed it. I think we should revert to `iterator(start, limit)` and then `BlueArtifactFactory.resolve()` should also take start/limit in to account, otherwise we will end up reading all artifacts. Even probably BlueArtifactFactory.getArtifacts() could be expensive. I think we should fix it before merging, so I am marking it as 🐛